### PR TITLE
[libjpeg-turbo] Add official CMake targets in usage

### LINKED ISSUE #30696: [libjpeg-turbo] find_package(JPEG REQUIRED) cannot find libturbojpeg.a
****

**Is your feature request related to a problem? Please describe.**
libjpeg-turbo package has 2 APIs: libjpeg.a and libturbojpeg.a
however, find_package(JPEG REQUIRED) in CMakeLists.txt can only find libjpeg.a
My program tries to use the TurboJPEG API and is not able to compile.
It needs to link to libturbojpeg.a.
The variable ${JPEG_LIBRARIES} only has libjpeg.a
See attached sample program.

**Proposed solution**
JPEG_LIBRARIES should contain both libjpeg.a and libturbojpeg.a

**Describe alternatives you've considered**
My current workaround is to change the following line in my CMakeLists.txt
target_link_libraries(testjpeg PRIVATE ${JPEG_LIBRARIES})
to
target_link_libraries(testjpeg PRIVATE /specific/vcpkg/path/to/libturbojpeg.a)

**Additional context**
[testjpegsrc.zip](https://github.com/microsoft/vcpkg/files/11171122/testjpegsrc.zip)

### PR DESCRIPTION
**Pull request by LilyWangLL**
Fixes #30696, add [official](https://github.com/libjpeg-turbo/libjpeg-turbo/commit/b8200c6601e130bd695773ed5f75ef57157aa505) CMake targets in `usage`. 
```
    find_package(libjpeg-turbo CONFIG REQUIRED)
    target_link_libraries(main PRIVATE libjpeg-turbo::jpeg libjpeg-turbo::turbojpeg)
```
Usage passed on the following triplets:
```
x64-windows
x64-linux
```

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] ~~SHA512s are updated for each updated download~~
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by Neumann-A** [comment]
NO!

**Comment by LilyWangLL** [comment]
> NO!

Why this wrapper could not be removed?

**Comment by Neumann-A** [comment]
Because the wrapper is fixing a builtin cmake module. 
Maybe you should look at what the wrapper is actually doing and understanding why it exists instead of simply removing it. 
From the discussion of the linked issue i also see no inherit reason for this pr.

**Comment by dg0yt** [comment]
You can try to install the `cmake-user` test port locally if you don't want to wait for CI.

**Comment by LilyWangLL** [comment]
I reverted my wrong changes, and add the official CMake targets provided by `libjpeg-turbo` to `usage`. Is this change possible?

> However, libjpeg-turbo installs a pkg-config module libturbojpeg. This is accessible via pkg_check_modules(...) from CMake's built-in FindPkgConfig.cmake in connection with a pkg-config program (e.g. from port pkgconf). But I don't know if the libturbojpeg module is well tested on Windows.

Do I need also provide the `pkg_check_modules` usage?

cc @dg0yt @Neumann-A

**Comment by JackBoosY** [comment]
Since the upstream provides export cmake targets, we should **MODIFY** wrapper file: remove all the content and call `find_package(.. CONFIG)`, then provide the origin macros to the current users to avoid compatibility issues.

The usage file can be removed since the module mode is deprecated.

**Comment by dg0yt** [comment]
> ince the upstream provides export cmake targets, we should **MODIFY** wrapper file: remove all the content and call `find_package(.. CONFIG)`, then provide the origin macros to the current users to avoid compatibility issues.
> 
> The usage file can be removed since the module mode is deprecated.
> 
> I think we have many samples in other ports now.

And we know the problems from that approach!

File: ports/libjpeg-turbo/usage [hunk 1]
```diff
@@ -3,3 +3,8 @@ The package libjpeg-turbo is compatible with built-in CMake targets:
     find_package(JPEG REQUIRED)
     target_link_libraries(main PRIVATE ${JPEG_LIBRARIES})
     target_include_directories(main PRIVATE ${JPEG_INCLUDE_DIR})
+
+libjpeg-turbo also provides CMake targets:
+
+    find_package(libjpeg-turbo CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE libjpeg-turbo::jpeg libjpeg-turbo::turbojpeg)
```
Comment (dg0yt) [inline ports/libjpeg-turbo/usage:10]: I would suggest an alternative:
~~~
libjpeg-turbo provides CMake targets for the TurboJPEG C API:

    find_package(libjpeg-turbo CONFIG REQUIRED)
    target_link_libraries(main PRIVATE libjpeg-turbo::turbojpeg)
~~~

**Review by dg0yt — COMMENTED** [review]
IMO config should only be advertised for the specific TurboJPEG usage.

The standard libjpg API is sufficiently covered by CMake's Find module. It is up to Kitware to decide for chainloading. (They do it for CURL.)

**Comment by JackBoosY** [comment]
> And we know the problems from that approach!

I don't understand. What else problem did we meet except the find module file provieded by cmake official?

**Comment by dg0yt** [comment]
- The canonical and widely used pattern for a libjpeg API *regardless of implementation* is:
  ~~~cmake
  find_package(JPEG)
  ~~~
- The implementation of this pattern is Kitware's  `FindJPEG.cmake`. It is maintained, not deprecated.
  It provides a target `JPEG::JPEG`, since CMake 3.12. This could be added to the `usage`, instead of variables.
- Find module variables are sometimes used to populate pkgconfig files. This pattern breaks when targets are injected.
- The wrapper in vcpkg also adds polyfill for older version of CMake (user side).

**Comment by Neumann-A** [comment]
..................

libjpeg-turbo provides two things:

a) jpeg api: If only use jpeg api and that could be fullfilled by another library use the JPEG::JPEG target provided by the cmake module (theoretically vcpkg could have a jpeg port which redirects to libjpeg-turbo or any other implementation of the jpeg api)
b) jpeg-turbo api: additional api provided by libjpeg-turbo accessed via the exported targets in config mode. 

As such the port should provide usage for a and b

**Comment by JackBoosY** [comment]
@Neumann-A How about this:
```cmake
list(REMOVE_ITEM ARGS "MODULE")
_find_package(libjpeg-turbo ${ARGS})

add_library(JPEG::JPEG IMPORTED)
set_target_properties(JPEG::JPEG PROPERTIES INTERFACE_LINK_LIBRARIES libjpeg-turbo <or any exported target>)

if (<HAS_FEATURE_1>)
    set_target_properties(JPEG::JPEG PROPERTIES INTERFACE_LINK_LIBRARIES libjpeg-turbo <or any exported target>)
endif()
...

set(JPEG_LIBRARIES libjpeg-turbo< or any exported target>)
get_target_property(JPEG_INCLUDE_DIR libjpeg-turbo< or any exported target> INTERFACE_INCLUDE_DIRECTORIES)
```

**Comment by dg0yt** [comment]
I don't get why we should touch a working port and wrapper.

File: ports/libjpeg-turbo/usage (hunk 1 above)
Comment (dg0yt) [reply ports/libjpeg-turbo/usage:10]: I would like to extend the proposal. Full new version:
~~~
libjpeg-turbo is compatible with built-in implementation-agnostic CMake targets:

    find_package(JPEG REQUIRED)
    target_include_directories(main PRIVATE JPEG::JPEG)

libjpeg-turbo provides CMake targets for the TurboJPEG C API:

    find_package(libjpeg-turbo CONFIG REQUIRED)
    target_link_libraries(main PRIVATE libjpeg-turbo::turbojpeg)
~~~

**Comment by JackBoosY** [comment]
> I don't get why we should touch a working port and wrapper.

Converting the module mode to the config mode can effectively solve the hidden dangers caused by the upgrade.

**Comment by Neumann-A** [comment]
> Converting the module mode to the config mode can effectively solve the hidden dangers caused by the upgrade.

No it just blocks vendored modules and can lead to more issues ;)

**Comment by JackBoosY** [comment]
> No it just blocks vendored modules and can lead to more issues ;)

There have been countless cases where we have these kinds of issues, so I think we can use this solution to proactively avoid them.
I also think that any other problems that arise from this can be solved by us.

**Comment by dg0yt** [comment]
> Converting the module mode to the config mode can effectively solve the hidden dangers caused by the upgrade.

Which upgrade?

> There have been countless cases where we have these kinds of issues.

Which JPEG problems?

If you want `FindJPEG.cmake` to chainload `libjpeg-turbo CONFIG`, please do that upstream at Kitware CMake, so that it doesn't need to be maintained in vcpkg. Maybe you also need to make `libjpeg-turbo CONFIG` provide the CMake variable interface of `JPEG` if module.

> >  Find module variables are sometimes used to populate pkgconfig files. This pattern breaks when targets are injected.
> 
> Do you have any example code in this case?

CURL (patch) and GDAL (upstream) carry CMake code to flatten CMake targets into lists of libs.
`vcpkg_fixup_pkgconfig` carries code to check for `::`, and getting this check merged required fixes to a number of ports.

File: ports/libjpeg-turbo/usage (hunk 1 above)
Comment (dg0yt) [reply ports/libjpeg-turbo/usage:10]: My proposal is now in https://github.com/microsoft/vcpkg/pull/31071.

**Comment by dg0yt** [comment]
Given that this PR already became a broader discussion:

In the not so far future we may need to sort out API alternatives here, but I don't know much about the details.

- AFAIU the port has incompatible libjpeg API alternatives as features:
  https://github.com/microsoft/vcpkg/blob/b3e0a1e625a0cc00073efb763e39088e5b75939e/ports/libjpeg-turbo/vcpkg.json#L18-L23
- There is now a beta of libjpeg-turbo 3.0, https://github.com/libjpeg-turbo/libjpeg-turbo/releases/tag/2.1.91.
  I assume there will be a significant interest in using the new version soon.

**Comment by LilyWangLL** [comment]
Closed this PR as duplicate of #31071.

**Pull request closed without merging**